### ISSUE TITLE
Separate progress for fine-tuning and inferencing 

### ISSUE THREAD

**Issue by Nateowami** [sillsdev/serval#477]
## Background

Currently a `TranslationBuild` has a `PercentCompleted` property that estimates the overall completion of finetuning + inferencing. I would like a separate progress for each step so that we can break them out separately for the user. 

## Proposal

Add these properties to the `TranslationBuild`:
- `public double? FineTuneProgress`
- `public double? InferenceProgress`

Both values should range from 0 to 1, and should be null when the associated process has not been started. You could pick other names, but I would ask that "percent" not be used in the name, because it is not actually a percentage (the misleading name of `PercentCompleted` resulted in progress being shown incorrectly for a short time).

**Comment by Enkidu93** [comment]
@johnml1135, this might be another good next issue for @mudiagaobrikisil (?).

**Comment by johnml1135** [comment]
Option 1 (better % splitting) with modification:
* When giving an update, also add the step number in this structured format:
  * <message about status>: 12345 of 23456
  * Also, on the build's executionData, the number of training segments and the number of pretranslation segments are available.

**Comment by pmachapman** [comment]
@johnml1135 Would you add a new property and have its type as this new object?

**Comment by johnml1135** [comment]
Here are the API level options that I can think of:
Option 1 - repurpose Build message
* Replace the "message" field getBuild with a Json string with lots of data
Option 2 - Add new object
* Add new fields to the Build object, detailing every desired statistic
Option 3 - extend "ExecutionData":
* A new field "ExecutionData" is being added (under review) that is a dict<string,sting>.  It could handle some or all of this data.
Option 4 - hybrid:
* Add most fields to `executionData`, but also create a new field `executionPhases` as an array similar to the one above.  This could hold just the static "this is the plan" data while `executionData` holds the current state in fields such as `currentPhaseIndex` and `currentStepInCurrentPhase` along with the other meta-data such as `estimatedEndTime`.

@pmachapman  - you will likely be writing the code to interpret and utilize the data.  What would be best for you?

**Comment by johnml1135** [comment]
@mudiagaobrikisil - you can start the calculation of the more accurate phases.  Please talk to @mshannon-sil about how to get setup with ClearML sessions to utilize a GPU - or on how to start the task of from locally.

**Comment by pmachapman** [comment]
@johnml1135 
Option 1 will break compatibility, as we already retrieve the message property.

I think options 2-4 are preferable - which ever is easier to implement or fits the existing architecture better. Perhaps the only problem with option 3 is that key names may change? (I'm a strongly typed kinda guy...)

Option 2 would be the easiest for me, as you would likely have a new version of the Serval.Client library anytime you add more fields to it, and these would be able to be documented in the code docs for the new properties, etc.

**Comment by ddaspit** [comment]
We should go with option 2.

**Comment by mudiagaobrikisil** [comment]
I want with option 2 as recommended. I have created new fields to the build object that capture the desired statistics. I am working on sending the actual numbers to ClearML and then fetching it from serval and populating the desired fields.

**Comment by johnml1135** [comment]
@mudiagaobrikisil  - I am a little confused.  We will need to have the ClearML job push scalar data (https://clear.ml/docs/latest/docs/guides/reporting/scalar_reporting/) to the ClearML server and then retrieve it from Serval to populate the fields.  Is that what you were referring to?

**Comment by mudiagaobrikisil** [comment]
@johnml1135 , yes. That's what I am referring to. Just didn't put it well in writing.

**Comment by ddaspit** [comment]
@mudiagaobrikisil Here is what the model might look like:

Build
- ProgressStages: list of ProgressStage objects

ProgressStage
- Name: string
- Step: integer
- Message: string (optional)
- PercentCompleted: double (optional)

The naming is consistent with the existing properties, but I'm open to different naming.

**Comment by Nateowami** [comment]
I would avoid using "Percent" in "PercentCompleted", as it implies a range of 0-100. We already had one bug caused by this.

**Comment by ddaspit** [comment]
"Percent" is ambiguous and can mean a number between 0-1 or 0-100, since they are equivalent. In this instance, we could use a number between 0-100, since that is how percent is probably more commonly used. It would make it inconsistent with the existing `PercentCompleted` property. Previously, John suggested using `TotalSteps` instead of `PercentCompleted`. The caller could compute the percent completed from `Step` and `TotalSteps`. This would require more internal Serval changes, since we don't currently pass the total steps from the ClearML job. We could also use a totally different name here, such as `Progress`.

**Comment by johnml1135** [comment]
Based upon the last conversation with Nathaniel, I believe we will implement "step" and "totalSteps" - and precent can be derived from that.

**Comment by Enkidu93** [comment]
> Based upon the last conversation with Nathaniel, I believe we will implement "step" and "totalSteps" - and precent can be derived from that.

@ddaspit, are `step` and `totalSteps` alright with you? (as opposed to two 'percent completes')

**Comment by Enkidu93** [comment]
I'll hold off on this issue until after the Wednesday Serval meeting. This might get pushed back.

**Comment by ddaspit** [comment]
It sounds like it is either

- `step` and `totalSteps`
- `progress`: value between 0-100

We could have both. We should double-check with the SF team before deciding. Whatever we do, the build DTO should be updated to be consistent and we should deprecate any old properties.
